specs/verge/beacon-chain.md

```diff
@@ -0,0 +1,208 @@
+# The Verge -- The Beacon Chain
+
+## Table of contents
+
+<!-- TOC -->
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Introduction](#introduction)
+- [Custom types](#custom-types)
+- [Preset](#preset)
+  - [Execution](#execution)
+- [Containers](#containers)
+  - [Extended containers](#extended-containers)
+    - [`ExecutionPayload`](#executionpayload)
+    - [`ExecutionPayloadHeader`](#executionpayloadheader)
+  - [New containers](#new-containers)
+    - [`SuffixStateDiff`](#suffixstatediff)
+    - [`StemStateDiff`](#stemstatediff)
+    - [`IPAProof`](#ipaproof)
+    - [`VerkleProof`](#verkleproof)
+    - [`ExecutionWitness`](#executionwitness)
+- [Beacon chain state transition function](#beacon-chain-state-transition-function)
+  - [Execution engine](#execution-engine)
+    - [`notify_new_payload`](#notify_new_payload)
+  - [Block processing](#block-processing)
+    - [Execution payload](#execution-payload)
+      - [`process_execution_payload`](#process_execution_payload)
+- [Testing](#testing)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Introduction
+
+This upgrade adds transaction execution to the beacon chain as part of the Verge upgrade.
+
+## Custom types
+
+| Name | SSZ equivalent | Description |
+| - | - | - |
+| `StateDiff` | `List[StemStateDiff, MAX_STEMS]` | Only valid if list is sorted by stems |
+| `BandersnatchGroupElement` | `Bytes32` | |
+| `BandersnatchFieldElement` | `Bytes32` | |
+| `Stem` | `Bytes31` | |
+
+## Preset
+
+### Execution
+
+| Name | Value |
+| - | - |
+| `MAX_STEMS` | `2**16` |
+| `MAX_COMMITMENTS_PER_STEM` | `33` |
+| `VERKLE_WIDTH` | `256` |
+| `IPA_PROOF_DEPTH` | `8` |
+
+## Containers
+
+### Extended containers
+
+#### `ExecutionPayload`
+
+```python
+class ExecutionPayload(Container):
+    # Execution block header fields
+    parent_hash: Hash32
+    fee_recipient: ExecutionAddress  # 'beneficiary' in the yellow paper
+    state_root: Bytes32
+    receipts_root: Bytes32
+    logs_bloom: ByteVector[BYTES_PER_LOGS_BLOOM]
+    prev_randao: Bytes32  # 'difficulty' in the yellow paper
+    block_number: uint64  # 'number' in the yellow paper
+    gas_limit: uint64
+    gas_used: uint64
+    timestamp: uint64
+    extra_data: ByteList[MAX_EXTRA_DATA_BYTES]
+    base_fee_per_gas: uint256
+    block_hash: Hash32  # Hash of execution block
+    # Extra payload field
+    execution_witness: ExecutionWitness
+    transactions: List[Transaction, MAX_TRANSACTIONS_PER_PAYLOAD]
```

Comment:
`execution_witness` should be after `transactions` to comply with `ExecutionPayloadHeader`

```suggestion
    # Extra payload field
    transactions: List[Transaction, MAX_TRANSACTIONS_PER_PAYLOAD]
    execution_witness: ExecutionWitness
```